### DISCUSSION TITLE
Update v6.5

### DISCUSSION THREAD

**Discussion by TBog** [TBog/TBLauncher#186] [Release]
* Replaced result `ListView` with `RecyclerView` - _better scroll handling_
* New Widget pages - _up, down, left, right_
* fix invisible dialog buttons

<hr /><em>This discussion was created from the release <a href='https://github.com/TBog/TBLauncher/releases/tag/v6.5'>Update v6.5</a>.</em>

**Comment by heiwiper** [comment]
Thanks for the new update, I've been looking forward to it.
I would like to confirm if this is an issue or I'm just not getting how the new pages work. So when I set the vertical pages to "Center page" and horizontal pages to "2 pages..." The left page becomes the Middle page and all widgets are created there, the popup menu gives the possibility to move the widget to the left page although that doesn't change anything, the right page isn't usable in this case.

**Reply by TBog** [reply]
You found an issue, I'll try to debug and fix it asap.

**Comment by ltguillaume** [comment]
Is there any way to prioritize keeping TBLauncher in memory somehow (programmatically I mean, I already disabled battery optimizations). Like with a foreground service? Currently, it just restarts almost every time I get back to it, and with v6.5 it still takes 30s at least, somtimes even 2 minutes (!) for the apps list to show anything (I'm just staring at the launcher button rotating and rotating and rotating), while it can take up to 3 more minutes for all icons to show up.

**Reply by TBog** [reply]
I'm not sure.
* Why does it take so long to load the app list? 
* Does the icon pack have a big xml? 
	- If you don't use an icon pack does it still take that long?
	- If while you wait for the app list to load you press the reload action (I do it pressing the 9 square button next to the search bar), does this speed up the load or no?

**Reply by ltguillaume** [reply]
Oh wow, I just realized that if I _don't_ choose a default icon pack, I can still handpick icons from that icon pack, which in effect allows me to do what I requested in an issue earlier: save _all_ icons in TBLauncher's database, instead of having it look them up in the icon pack every time.

- Yes, Whicons has a ton of icons, so its XML is probably one of the largest of all icon packs. But I'm happy to know now I can get around that.
  - Yes, let's see what happens if I pick custom icons for all apps while setting the icon pack to "System".
  - I have tried this before, it did help when apps didn't show up at all, but icons from the icon pack would often still be absent (for minutes, or sometimes they wouldn't show up at all anymore). I suppose there's definitely still something to gain in icon pack handling if I compare this with e.g. Lawnchair(-based launchers), which were much faster in both listing the apps and in presenting Whicons on the same device, also after a force close.

**Reply by TBog** [reply]
> I suppose there's definitely still something to gain in icon pack handling if I compare this with e.g. Lawnchair(-based launchers), which were much faster in both listing the apps and in presenting Whicons on the same device

1. When I have time I'll look into optimizing the icon pack parsing.
2. I'll also add a cache in the DB with the icons from the current icon pack.

**Reply by ltguillaume** [reply]
Sounds great!

I guess in the upcoming days I will see if such a cache will resolve the issues I'm having, because I suspect I now effectively have the same thing by choosing all icons manually.

**Reply by ltguillaume** [reply]
The difference between using Whicons as default icon pack and System (with all icons set manually) is 20-30s before apps show up after closing the app via settings, vs. 4s. As you can imagine, I am very happy now 😉

**Reply by TBog** [reply]
Are you telling me to implement the icon pack cache? 😁

The only problem I have is that I can't determine when the cache is outdated. And I don't want to build the cache each time the launcher starts / finishes to load the icon pack.

BTW thanks for the information you provided.

**Reply by TBog** [reply]
Also now I'll have to implement a viewer for the custom icons and cached app icons to debug, but it will be useful anyways.

**Reply by ltguillaume** [reply]
> Are you telling me to implement the icon pack cache? 😁

Haha, not telling you to do anything, just providing info 😛 And basically I've got my own icon cache already now, it just took a few minutes to manually select the icon pack icons for some apps.

> The only problem I have is that I can't determine when the cache is outdated.

I'm not sure what you mean there: do you mean "I don't have a way of knowing when icons in the pack are updated"? Personally I wouldn't mind, I'd like the icons to stay the same and I would prefer a function to rebuild the cache manually, if at all.

Or do you mean that the cache could be outdated because of app (un)installs?

> Also now I'll have to implement a viewer for the custom icons and cached app icons to debug, but it will be useful anyways.

If you choose "Change icon", you already see the current (cached) icon at the bottom, right? So what do you mean exactly?

**Reply by TBog** [reply]
> > Are you telling me to implement the icon pack cache? 😁
> 
> Haha, not telling you to do anything, just providing info 😛 And basically I've got my own icon cache already now, it just took a few minutes to manually select the icon pack icons for some apps.

Sorry for the bad UX (user experience) 

> > The only problem I have is that I can't determine when the cache is outdated.
> 
> I'm not sure what you mean there: do you mean "I don't have a way of knowing when icons in the pack are updated"? Personally I wouldn't mind, I'd like the icons to stay the same and I would prefer a function to rebuild the cache manually, if at all.

Exactly.
I get that you don't mind having a rebuild cache button, but is this the way other launchers do it? Is a new user going to figure out why the icon is not updated? 

> > Also now I'll have to implement a viewer for the custom icons and cached app icons to debug, but it will be useful anyways.
> 
> If you choose "Change icon", you already see the current (cached) icon at the bottom, right? So what do you mean exactly?

But you have to do that for every app.
What if you want to revert all app icons to the system ones? 
It's also why I did the `tags manager`; to be able to rename a tag without going through all the apps that had that tag and remove the old tag then add the new tag.

**Reply by ltguillaume** [reply]
> Sorry for the bad UX (user experience)

Well it does what I want now, I just wish I knew this before 😛

> But is this the way other launchers do it?

Not sure, but I think you're right that users would expect updated icons in the launcher when the icons are updated in the pack.

I suppose the best way to not update icons every time is to keep track of the versioncode of the update pack(s) and rebuild the icon cache whenever the _chosen default_ update pack has been updated.

>  What if you want to revert all app icons to the system ones?

I don't know if such a use case is prevalent enough to require building a whole viewer, but that's up to you of course. I think `Clear icon cache` would suffice, and instead of that manual `Rebuild icon cache` I proposed earlier, a versioncode check of the icon pack + rebuild  would be better.

**Reply by TBog** [reply]
The version code check of the default icon pack is already implemented on de `develop` branch. I still need more testing and to clear the cache when the icon pack changes, right now if the new icon pack has the same version (even if it's a different application) the cache is not rebuilt. Oops.

**Reply by ltguillaume** [reply]
> The version code check of the default icon pack is already implemented on de develop branch. 

...and here I was thinking I'm smart and all

**Comment by heiwiper** [comment]
The keyboard doesn't seem to follow the search bar on the first tap which requires tapping twice in order to get the keyboard to start following the search bar.

**Reply by TBog** [reply]
Sorry for missing your comment here, I'm working on this.

**Comment by ltguillaume** [comment]
Hehe, I see the `Add tags` dialog still has hardcoded green lines around the selected tags (and a green `+` button), but the `Rename` dialog has been improved 🙂
![image](https://user-images.githubusercontent.com/4777345/130335260-845df616-0376-4683-91aa-7ca8655a3dd3.png)

Saw a new strange thing in the `Close launcher` dialog:
![image](https://user-images.githubusercontent.com/4777345/130335203-d1290764-82fa-4bd2-a9ac-e0328fa2ed67.png)

**Reply by TBog** [reply]
> Saw a new strange thing in the `Close launcher` dialog

fixed on the `develop` branch, commit 421788bece41f56ca346a1b49777f3b401206ce0

**Reply by TBog** [reply]
The `add tags` dialog is going to change after #55

**Reply by ltguillaume** [reply]
And this?

![image](https://user-images.githubusercontent.com/4777345/131725577-6279d49e-8e93-48bb-aa0b-815ea1c7e568.png)

**Reply by TBog** [reply]
Forgot about the shortcut confirmation activity. I'll fix it when I get back.

**Reply by TBog** [reply]
fixed shortcut confirm invisible buttons in PR #198
